### PR TITLE
updates to `customizeHLTforPatatrack` to limit overwriting of HLT modules

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR updates `customizeHLTforPatatrack` to limit cases where the customisation overwrites modules that may already exist in the HLT menu (e.g. the `cpu` 'branch' of certain SwitchProducers).

*Reminder* : currently, the HLT menu in `cmssw:master` contains the CPU part of the Patatrack developments (in other words, it contains SwitchProducers that only have a `cpu` branch, not a `cuda` one).

With this update, if a certain object exists already, it is not overwritten anymore, giving precedence to what is in the input configuration (usually, the HLT menu from `ConfDB`).

In particular, this resolves an action item from [CMSHLT-2187](https://its.cern.ch/jira/projects/CMSHLT/issues/CMSHLT-2187), regarding the value of the parameter `hltPixelTracksSoA.cpu.idealConditions`; the latter is set to `False` in `ConfDB`, but is currently changed to `True` in `customizeHLTforPatatrack`; with this PR, the parameter remains as in the HLT menu if it is already present (i.e. `False` with the latest menus).

The PR also includes
 - an addition to `customisePixelGainForRun2Input`, to ensure that the parameter `(SiPixelRawToClusterCUDA).isRun2` is set to `True` when Run-2 inputs are used (this would be needed once instances of `SiPixelRawToClusterCUDA` will come directly from the HLT config in `ConfDB`, with `isRun2` set to `False`);
 - a bugfix for the handling of `process.hltEcalRecHit.cuda` (the bug was introduced in #36536).

Changes in outputs are expected for wfs where `customizeHLTforPatatrack` is applied to the `GRun` menu, e.g. wf `11634.{506,512,522}` (for example, because of `hltPixelTracksSoA.cpu.idealConditions` changing from `True` to `False`).

Profited from suggestions by @fwyzard.

FYI: @silviodonato

#### PR validation:

`runTheMatrix.py -w gpu -l 11634.506,11634.512,11634.522`, plus manual tests with `hltGetConfiguration [..] --customise [..]`.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR:

N/A

### CODE REVIEW

**Comment by missirol** [comment]
type bugfix

> a bugfix for the handling of `process.hltEcalRecHit.cuda` (the bug was introduced in #36536).

File: HLTrigger/Configuration/python/customizeHLTforPatatrack.py [hunk 1]
```diff
@@ -540,46 +650,69 @@ def customiseEcalLocalReconstruction(process):
...
         )
 
+    elif not hasattr(process.hltEcalRecHit, 'cpu'):
+        raise Exception('unsupported configuration: "process.hltEcalRecHit" is a SwitchProducerCUDA, but does not have a "cpu" branch')
+
+    if not hasattr(process.hltEcalRecHit, 'cuda'):
+        process.hltEcalRecHit.cuda = process.hltEcalRecHit.cpu.clone(
+            triggerPrimitiveDigiCollection = 'unused'
+        )
+
```
Comment (missirol) [inline HLTrigger/Configuration/python/customizeHLTforPatatrack.py:663]: > a bugfix for the handling of `process.hltEcalRecHit.cuda` (the bug was introduced in #36536).

After #36536, `process.hltEcalRecHit.cuda` would be added only if `process.hltEcalRecHit` was not a SwitchProducer in the input configuration. This condition is false in the latest menus, so for those `process.hltEcalRecHit.cuda` was not being added.

File: HLTrigger/Configuration/python/customizeHLTforPatatrack.py [hunk 2]
```diff
@@ -286,112 +316,151 @@ def customisePixelTrackReconstruction(process):
...
+            )
+        )
+
+    elif hasattr(process.hltPixelTracksSoA, 'cpu'):
+        # if cpu branch of SwitchProducerCUDA exists, take hltPixelTracksCUDA (gpu)
+        # from hltPixelTracksSoA.cpu (cpu) to enforce same configuration parameters
+        process.hltPixelTracksCUDA = process.hltPixelTracksSoA.cpu.clone(
+            pixelRecHitSrc = "hltSiPixelRecHitsCUDA",
+            onGPU = True
+        )
```
Comment (missirol) [inline HLTrigger/Configuration/python/customizeHLTforPatatrack.py:353]: This relates to the treatment of the value of `process.hltPixelTracksSoA.idealConditions`. If `process.hltPixelTracksSoA.cuda` exists, it is not modified, and the GPU module (`hltPixelTracksCUDA`) is configured based on `process.hltPixelTracksSoA.cuda`.

**Comment by fwyzard** [comment]
You could define some helper functions to simplify the customisation:

```python
def load_if_mising(process, label, config):
if label not in process.__dict__:
    process.load(config)

def load_if_missing_and_clone(_process, _config, _label, _alias = None, **kwargs):
    if _alias is None:
        _alias = _label
    if _alias not in _process.__dict__:
        module = __import__(_config, globals(), locals(), [ _label ], 0)
        setattr(_process, _alias, getattr(module, _label).clone(**kwargs)
```

to be used as
```python
# these should be used only on GPUs, will crash otherwise
load_if_mising(process, 'siPixelGainCalibrationForHLTGPU', 'CalibTracker.SiPixelESProducers.siPixelGainCalibrationForHLTGPU_cfi')
load_if_mising(process, 'siPixelROCsStatusAndMappingWrapperESProducer', 'CalibTracker.SiPixelESProducers.siPixelROCsStatusAndMappingWrapperESProducer_cfi')
...

# transfer the beamspot to the gpu
load_if_missing_and_clone(process, 'RecoVertex.BeamSpotProducer.offlineBeamSpotToCUDA_cfi', 'offlineBeamSpotToCUDA', 'hltOnlineBeamSpotToCUDA',
    src = "hltOnlineBeamSpot"
)
```

Please note that I did not test that they actually work... and maybe you can find more compact names :-)

I didn't try to come up with something that support a `SwitchProducer`, but I can look into it if needed.

**Comment by missirol** [comment]
Thanks for the suggestion, @fwyzard . I'm trying to implement it.

What do you think about the following?
```py
# Utility to load file "config" in the process,
# if the process does not already hold a module named "label"
# Example:
#   _load_if_missing(process, 'SomeESProducer', 'SubSystem.Package.SomeESProducer_cfi')
def _load_if_missing(process, label, config, check = True):
    if label not in process.__dict__:
        process.load(config)
    if check and label not in process.__dict__:
        raise Exception('process does not have a module labelled "'+label+'" after loading "'+config+'"')

# Utility to add "member" to "obj" (if "member" is missing),
# using a clone of module "label" from file "config"
# (if "label" is not specified, it is set equal to "member")
# Example:
#   _clone_if_missing(process, 'hltMyFilter', 'SubSystem.Package.hltSomeFilter_cfi', 'hltSomeFilter', src = 'mySrc')
def _clone_if_missing(obj, member, config, label = None, **kwargs):
    if label is None:
        label = member
    if not hasattr(obj, member):
        module = __import__(config, globals(), locals(), [ label ], 0)
        setattr(obj, member, getattr(module, label).clone(**kwargs))
```
The last function is slightly changed wrt to the original suggestion (also, from your snippet I didn't grasp if starting the function arguments with `_` is important). It would allow to do both
```py
_clone_if_missing(process, 'hltOnlineBeamSpotToCUDA', 'RecoVertex.BeamSpotProducer.offlineBeamSpotToCUDA_cfi', 'offlineBeamSpotToCUDA',
    src = "hltOnlineBeamSpot"
)
```
and
```py
_clone_if_missing(process.hltSiPixelDigis, 'cuda', 'EventFilter.SiPixelRawToDigi.siPixelDigiErrorsFromSoA_cfi', 'siPixelDigiErrorsFromSoA',
    digiErrorSoASrc = "hltSiPixelDigiErrorsSoA",
    UsePhase1 = True
)
```
.. so, one part of the SwitchProducer customisations.

**Comment by fwyzard** [comment]
> from your snippet I didn't grasp if starting the function arguments with `_` is important

that's to make it less likely that a function argument would clash with an argument intended for `clone()`

**Comment by fwyzard** [comment]
I'd suggest 
```
def _clone_if_missing(obj, label, config, original_label = None, **kwargs):
```
to make the naming more consistent among the two functions.

**Comment by missirol** [comment]
Sorry for the multiple pushes; it should be done now. Hopefully the different commits make it a bit easier to review; I plan to eventually squash this into 2 commits (one for `customisePixelGainForRun2Input` and one for `customizeHLTforPatatrack`) before it is integrated.

**Comment by missirol** [comment]
The latest push only squashes these updates into 2 commits (no changes wrt last week). I did more manual checks on this latest version (comparing configs w/ and w/o the customisation, and comparing the configs-with-customisation w/ and w/o this PR), and I could not find any issues. The resulting configs look as expected.

I have one doubt about the testing. If I remember correctly, "enable gpu" will run additional wfs that use `customiseHLTforPatatrack` on machines with a GPU (so testing the GPU branches of the SwitchProducers); if so, I don't know how to ask the bot to run the same additional wfs on a machine without a GPU (to test the CPU branches of the SwitchProducers). I tested locally "runTheMatrix.py -w gpu -l 11634.506,11634.512,11634.522" without a GPU, and that went okay.

**Comment by missirol** [comment]
enable gpu

**Comment by missirol** [comment]
please test

**Comment by missirol** [comment]
In my understanding, the outputs of the PR tests look as expected.

 - Non-GPU tests show no differences as expected; by construction, this PR does not affect them.

 - For `11634.{512,522}` there are differences in the HLT outputs (e.g. "Trigger Results" for the HLT step); differences are expected since these wfs apply `customiseHLTforPatatrack` to the HLT step: pre-PR this means using HLT pixel quadruplets (the customisation overwrites what is in the menu), while post-PR HLT pixel triplets are used instead (because this is what is in the current HLT, and the updated customisation does not overwrite that).

 - For `11634.506`, there are no differences in the HLT triggerResults; this might be explained by the fact that this wf uses `customiseHLTforPatatrack+enablePatatrackPixelTriplets`, so HLT pixel triplets are used both pre- and post-PR. There are still some expected changes in the config pre- and post-PR (e.g. the `idealConditions` parameter of the HLT pixel-track reco), but this doesn't create a visible differences in the outputs (maybe because of the limited stats and the fact that the wf is without PU).

 - The many very-small differences in the RECO outputs of `11634.506` are expected, as I understand that those outputs are not fully reproducible (by construction, this PR does not affect any RECO outputs)
   https://github.com/cms-sw/cmssw/pull/35858#issuecomment-955201296
   https://github.com/cms-sw/cmssw/pull/36714#issuecomment-1014289704

Unless there are comments, or I find issues, I plan to sign this in the next few days.

**Comment by missirol** [comment]
please test

Rebased, and improved handling of parameters for timing of ECAL RecHits (thanks to suggestions from Silvio and Andrea M.), in view of the upcoming updates in [CMSHLT-2211](https://its.cern.ch/jira/projects/CMSHLT/issues/CMSHLT-2211).

@amassiro (cc: @fwyzard, @silviodonato), please have a look at 579db0b098185dd7d35b21bba0a6f9a0a1b76db9, and let me know if you have comments/corrections. Note that, in this customisation, `hltEcalRecHit.cuda` is cloned from `hltEcalRecHit.cpu`, so there is no need to edit `hltEcalRecHit.cuda.skipTimeCalib` explicitly.

**Comment by missirol** [comment]
please test

re-running the tests, as the latest IBs include updates to the HLT menu (wrt the last tests of this PR)

**Comment by missirol** [comment]
+hlt

 - Improvement of the GPU customisation functions in order to (1) avoid overwriting the content of the HLT menu, and (2) better ensure consistent configurations of CPU and GPU modules.
 - The PR also fixes a small bug, see https://github.com/cms-sw/cmssw/pull/36721#discussion_r785858990.
 - Differences in (GPU) wfs using this customisation function are expected and observed, see https://github.com/cms-sw/cmssw/pull/36721#issuecomment-1029166619.
 - TSG would like to have this PR in `CMSSW_12_3_0_pre5`; this customisation function is used (among other things) for timing studies.

**Pull request merged**